erts/emulator/beam/erl_bif_persistent.c

```diff
@@ -0,0 +1,816 @@
+     * sharing-preserving functions. However, literals must
+     * be copied in case the module holding them are unloaded.
+     */
+    INITIALIZE_SHCOPY(info);
+    info.copy_literals = 1;
+    term_size = copy_shared_calculate(tuple, &info);
+    ERTS_INIT_OFF_HEAP(&code_off_heap);
+    lit_area_size = ERTS_LITERAL_AREA_ALLOC_SIZE(term_size);
+    literal_area = erts_alloc(ERTS_ALC_T_LITERAL, lit_area_size);
+    ptr = &literal_area->start[0];
+    literal_area->end = ptr + term_size;
+    tuple = copy_shared_perform(tuple, term_size, &info, &ptr, &code_off_heap);
+    ASSERT(tuple_val(tuple) == literal_area->start);
+    literal_area->off_heap = code_off_heap.first;
+    DESTROY_SHCOPY(info);
+    hash_table->term[entry_index] = tuple;
+
+    erts_schedule_thr_prgr_later_op(table_updater, hash_table, &thr_prog_op);
+    suspend_updater(BIF_P);
+    ERTS_BIF_YIELD_RETURN(BIF_P, am_ok);
+}
+
+BIF_RETTYPE persistent_term_get_0(BIF_ALIST_0)
+{
+    HashTable* hash_table = (HashTable *) erts_atomic_read_nob(&the_hash_table);
+    Uint trap_data_size = (sizeof(TrapData)+sizeof(Eterm)-1) / sizeof(Eterm);
+    Eterm* hp = HAlloc(BIF_P, trap_data_size);
+    TrapData* trap_data;
+    Eterm res = NIL;
+
+    trap_data = (TrapData *) hp;
+    trap_data->header = make_pos_bignum_header(trap_data_size-1);
+    trap_data->table = hash_table;
+    trap_data->idx = 0;
+    trap_data->remaining = hash_table->num_entries;
+    res = do_get_all(BIF_P, trap_data, res);
+    if (trap_data->remaining == 0) {
+        BIF_RET(res);
+    } else {
+        /*
+         * Increment the ref counter to prevent an update operation (by put/2
+         * or erase/1) to delete this hash table.
+         */
+        erts_atomic_inc_nob(&hash_table->refc);
+        BIF_TRAP2(&persistent_term_get_all_export, BIF_P, make_boxed(hp), res);
+    }
+}
+
+BIF_RETTYPE persistent_term_get_1(BIF_ALIST_1)
+{
+    Eterm key = BIF_ARG_1;
+    HashTable* hash_table = (HashTable *) erts_atomic_read_nob(&the_hash_table);
+    Uint entry_index;
+    Eterm term;
+
+    entry_index = lookup(hash_table, key);
+    term = hash_table->term[entry_index];
+    if (is_boxed(term)) {
+        ASSERT(is_tuple_arity(term, 2));
+        BIF_RET(tuple_val(term)[2]);
+    }
+    BIF_ERROR(BIF_P, BADARG);
+}
+
+BIF_RETTYPE persistent_term_erase_1(BIF_ALIST_1)
+{
+    Eterm key = BIF_ARG_1;
+    HashTable* old_table;
+    HashTable* new_table;
+    Uint entry_index;
+    Eterm old_term;
+
+    if (!try_seize_update_permission(BIF_P)) {
+	ERTS_BIF_YIELD1(bif_export[BIF_persistent_term_erase_1],
+                        BIF_P, BIF_ARG_1);
+    }
+
+    old_table = (HashTable *) erts_atomic_read_nob(&the_hash_table);
+    entry_index = lookup(old_table, key);
+    old_term = old_table->term[entry_index];
+    if (is_boxed(old_term)) {
+        Uint new_size;
+
+        ASSERT(is_tuple_arity(old_term, 2));
+        old_table->term[entry_index] = NIL;
+        old_table->num_entries--;
```

Comment:
Here in erase/1 you write into the active `old_table`. I think this can lead to all kinds of buggy behavior with concurrent readers like get/0.
Instead treat `old_table` as read only, copy it and make changes in `new_table`.

Reply:
Fixed.